#' get Differentially Expressed Genes
#'
#' screens the \code{expression_data}, then merges it with the \code{regressor_matrix} and performs regressions to determine which gene show differential expression beyond \code{foldThreshDEG}.
#' can adjust for covariates using \code{cov}.
#'
#' @param expression_data
#' @param exp_symbol_col
#' @param regressor_matrix
#' @param reg_id_col
#' @param foldThreshDEG the X-fold expression threshold to be exceeded for a given gene to be
#' considered differentially expressed.
#' @param screenSD minimum gene expression SD; genes that show expression variability below
#' threshold are excluded.
#' @param verbose
#' @return  returns an object of class "DEG"
#'
#' An object of class "DEG" is a list containing:
#' \enumerate{
#'   \item a data frame containing gene symbols, raw regression coefficients, and a DEG variable;
#'   DEG values indicate whether a given gene showed expression values beyond threshold (1 = increased; -1 = decreased; 0 = below threshold).
#'   \item a dataframe generated by transposing the expression_data and merging with it with the regressor_matrix.
#'   \item a list of input arguments.
#' }
#' @examples
#' \dontrun{
#' #load example data
#' data("TAU_Trials3_Gene_CPM_Log2")
#' data("TAUTrials2022BC_Intervention_Rm1BadQC_RmB14")
#'
#' #get DEG
#' DEG_result <- get_DEG(expression_data = TAU_Trials3_Gene_CPM_Log2,
#' exp_symbol_col = 1,
#' regressor_matrix = TAUTrials2022BC_Intervention_Rm1BadQC_RmB14,
#' reg_id_col = 1,
#' foldThreshDEG = 2,
#' screenSD = 0,
#' verbose = TRUE)
#'
#' #view results
#' table(DEG_result$df_DEG$DEG)
#' }
#' @export
get_DEG <- function(expression_data,
                    exp_symbol_col = 1,
                    regressor_matrix,
                    reg_id_col = 1,
                    foldThreshDEG = 1.5,
                    screenSD = 0,
                    verbose = FALSE){

  #
  results <- list(df_DEG = NULL,
                  analysis_data = NULL,
                  inputs = as.list(environment()))
  class(results) <- "DEG"

  #gene screen
  expression_data_screened <- geneScreen(expression_data, exp_symbol_col, screenSD, verbose)

  #merge expression data with the predictor matrix
  analysis_data <- pre_DEG_merge(expression_data_screened, regressor_matrix, exp_symbol_col, reg_id_col)
  results$analysis_data <- analysis_data

  #get DEG
  df_DEG <- get_df_DEG(x = get_x_cov(analysis_data, regressor_matrix)$x,
                       cov = get_x_cov(analysis_data, regressor_matrix)$cov,
                       genes = analysis_data[,-c(1:ncol(regressor_matrix))],
                       foldThreshDEG = foldThreshDEG)

  results$df_DEG <- df_DEG

  if(verbose){
    print(paste0(table(df_DEG$DEG)[names(table(df_DEG$DEG)) == 1], " up-regulated genes identified"))
    print(paste0(table(df_DEG$DEG)[names(table(df_DEG$DEG)) == -1], " down-regulated genes identified"))
  }

  return(results)
}
