#' Transcript Origin Analysis
#'
#' `toa` performs transcript origin analysis
#'
#' @param x A numeric vector of values to be used as the predictor variable of interest.
#' @param genes A numeric matrix if gene expression values, where columns are individual genes.
#' @param toa_ref A data frame containing gene symbols and diagnosticity scores; should be generated by using `get_toa_ref`
#' @param cov A numeric vector/matrix of values, where columns are individual covariate measures; if NULL, unadjusted regression coefficients (dif) are produced
#' @param foldThreshDEG the X-fold expression threshold to be exceeded for a given gene to be considered differentially expressed.
#' @return A list object containing 2 data frames, including a summary table (df.means) and a dataframe containing individual gene symbol, regression coefficients, DEG, and toa_ref matching
#' @examples
#' #load example data
#' data("Chang")
#' data("toa_ref_epith_mesen")
#' #toa
#' toa_result <- toa(x = Chang$stress, genes = subset(Chang, select = -stress), toa_ref = toa_ref, foldThreshDEG = 1.25)
#' @export
toa <- function(x, genes, toa_ref, cov = NULL, foldThreshDEG = 1.5){

  #instantiate results list
  results <- list(df.means = NULL,
                  df.DEG = NULL,
                  inputs = as.list(environment()))

  #
  if(!valid_input_x(x) | !valid_input_genes(genes) | !valid_input_cov(cov) | !valid_input_toa_ref(toa_ref)){
    warning("invalid inputs; check warnings")
    return(results)
  }

  #get differentially expressed genes
  df.DEG = get_DEG(x, genes, cov, foldThreshDEG)

  if(is.null(df.DEG)){
    warning("ERROR: get_DEG failed")
    return(results)
  }

  #identify gene symbols that are present in both the analyzed and reference datasets, and that are up/down regulated
  df.DEG$ref_matched <- ifelse(df.DEG$gene %in% toa_ref$gene, 1, 0)
  df.DEG$ref_matched_expressed <- ifelse(df.DEG$ref_matched == 1 & df.DEG$DEG != 0, 1, 0)

  #compute mean diagnosticity scores (if any up/down regulated genes could be matched up to the reference data)
  if(sum(df.DEG$ref_matched_expressed) > 0){
    df.means <- get_df_means(df.DEG)
  }else{
    print("ERROR: could not aggregate diagnosticity score across up and down regulated genes matched to the reference dataset; check how many genes are ref-matched and up/down regulated")
    return(results)
  }

  #if no failure up to this point, update the results list with computed values
  results$df.means = df.means
  results$df.DEG = df.DEG

  return(results)
}
