#' Transcript Origin Analysis
#'
#' performs transcript origin analysis without bootstrapped estimates of mean diagnosticity scores.
#'
#' @inheritParams get_DEG
#' @param toa_ref a data frame containing gene symbols and diagnosticity scores; should be generated by using \code{get_toa_ref()}.
#' @return toa returns an object of class "toa"
#'
#' An object of class "toa" is a list containing:
#' \enumerate{
#'   \item a results data frame.
#'   \item a raw coefficients data frame.
#'   \item a list of input arguments.
#' }
#' @examples
#' #load example data
#' data("Chang")
#' data("epith_mesen_ref_raw")
#' #get diagnosticity scores
#' toa_ref_epith_mesen <- get_toa_ref(gene_symbols = epith_mesen_ref_raw[,1],
#'                                    exp_treatment = epith_mesen_ref_raw[,2:11],
#'                                    exp_control = epith_mesen_ref_raw[,12:21])
#' #toa
#' toa_result <- toa(x = Chang$stress,
#'                   genes = subset(Chang, select = -stress),
#'                   toa_ref = toa_ref_epith_mesen,
#'                   foldThreshDEG = 1.25)
#' @export
toa <- function(x, genes, toa_ref, cov = NULL, foldThreshDEG = 1.5){

  #instantiate toa result object
  results <- list(df_results = NULL,
                  df_DEG = NULL,
                  inputs = as.list(environment()))
  class(results) <- "toa"

  #check inputs
  if(!valid_input_x(x) | !valid_input_genes(genes) | !valid_input_cov(cov) | !valid_input_toa_ref(toa_ref)){
    warning("invalid inputs; check warnings")
    return(results)
  }

  #get differentially expressed genes
  df_DEG = get_DEG(x, genes, cov, foldThreshDEG)

  if(is.null(df_DEG)){
    warning("ERROR: get_DEG failed")
    return(results)
  }

  #identify gene symbols that are present in both the analyzed and reference datasets, and that are up/down regulated
  df_DEG$ref_matched <- ifelse(df_DEG$gene %in% toa_ref$gene, 1, 0)
  df_DEG$ref_matched_expressed <- ifelse(df_DEG$ref_matched == 1 & df_DEG$DEG != 0, 1, 0)

  #merge diagnosticity scores
  df_DEG <- merge(df_DEG, toa_ref, by = "gene", all.x = TRUE)

  #compute mean diagnosticity scores (if any up/down regulated genes could be matched up to the reference data)
  if(sum(df_DEG$ref_matched_expressed) > 0){
    df.means <- get_df_means(df_DEG)
  }else{
    print("ERROR: could not aggregate diagnosticity score across up and down regulated genes matched to the reference dataset; check how many genes are ref-matched and up/down regulated")
    return(results)
  }

  #if no failure up to this point, update the results list with computed values
  results$df_results = df.means
  results$df_DEG = df_DEG

  return(results)
}
